### ISSUE TITLE
Module '"@prisma/client"' has no exported member 'PrismaClient'.ts(2305)

### ISSUE THREAD

**Issue by kasir-barati** [prisma/prisma#13672]
## I tried to use **pnpm** in my proj but it seems it does not work

I see this two issues too but none of them helped:
- https://github.com/prisma/prisma/issues/2584
- https://github.com/prisma/prisma/issues/7484

### How to reproduce

1. Clone the repo: https://github.com/kasir-barati/basic-prisma-nestjs-restful-api
2. `pnpm install`
3. `pnpm prisma generate`
4. open `src/shared/modules/prisma/prisma.service.ts`
5. See error
![Screenshot](https://user-images.githubusercontent.com/73785723/172387670-f67bc805-5307-43ca-a07c-b0902628c60e.png)

### Prisma information

See repo

### Environment & setup

- OS:
```cmd
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 22.04 LTS
Release:        22.04
Codename:       jammy
```
- Database: PostgreSQL
- Node.js version: v16.15.0

### Prisma Version

```cmd
$ pnpm prisma -v
prisma                  : 3.14.0
@prisma/client          : 3.14.0
Current platform        : debian-openssl-1.1.x
Query Engine (Node-API) : libquery-engine 2b0c12756921c891fec4f68d9444e18c7d5d4a6a (at node_modules/.pnpm/@prisma+engines@3.14.0-36.2b0c12756921c891fec4f68d9444e18c7d5d4a6a/node_modules/@prisma/engines/libquery_engine-debian-openssl-1.1.x.so.node)
Migration Engine        : migration-engine-cli 2b0c12756921c891fec4f68d9444e18c7d5d4a6a (at node_modules/.pnpm/@prisma+engines@3.14.0-36.2b0c12756921c891fec4f68d9444e18c7d5d4a6a/node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x)
Introspection Engine    : introspection-core 2b0c12756921c891fec4f68d9444e18c7d5d4a6a (at node_modules/.pnpm/@prisma+engines@3.14.0-36.2b0c12756921c891fec4f68d9444e18c7d5d4a6a/node_modules/@prisma/engines/introspection-engine-debian-openssl-1.1.x)
Format Binary           : prisma-fmt 2b0c12756921c891fec4f68d9444e18c7d5d4a6a (at node_modules/.pnpm/@prisma+engines@3.14.0-36.2b0c12756921c891fec4f68d9444e18c7d5d4a6a/node_modules/@prisma/engines/prisma-fmt-debian-openssl-1.1.x)
Default Engines Hash    : 2b0c12756921c891fec4f68d9444e18c7d5d4a6a
Studio                  : 0.460.0
```

**Comment by kasir-barati** [comment]
> Which version of pnpm are you using? (might help to reproduce)

@Jolg42 
```cmd
$ pnpm -v
6.32.3
```

**Comment by kasir-barati** [comment]
@Jolg42 After Upgrading to latest version - ATM latest version is `7.1.9` - the problem does not resolved.

**Comment by kasir-barati** [comment]
@Jolg42 OK but I have the error. I guess you are on mac :confused: 
BTW I have this problem.

**Comment by richardwardza** [comment]
Hi @Jolg42 

I have the same issue with this repository I'm working on: https://github.com/richardwardza/prisma-lambda-layers
If I use `pnpm` then I also get `Module '"@prisma/client"' has no exported member 'PrismaClient'.ts(2305)` on my `import { PrismaClient } from '@prisma/client';`

This is after the `prisma generate` has run:
```
src/layers/prisma postinstall$ prisma generate
│ Environment variables loaded from .env
│ Prisma schema loaded from prisma/schema.prisma
│ ✔ Generated Prisma Client (3.15.1 | library) to ./../../../node_modules/.pnpm/@prisma+client@3.15.1_prisma@3.15.1/node_modules/@prisma/client in 385ms
│ You can now start using Prisma Client in your code. Reference: https://pris.ly/d/client
│ ```
│ import { PrismaClient } from '@prisma/client'
│ const prisma = new PrismaClient()
│ ```
```

The `node_modules/.pnpm/@prisma+client@3.15.1_prisma@3.15.1/node_modules/@prisma/client` directory does exist and has the following:

```
richard:bison hello > l node_modules/.pnpm/@prisma+client@3.15.1_prisma@3.15.1/node_modules/@prisma/client
total 88
-rw-r--r--   4 richard  staff  11357 Jun  9 21:14 LICENSE
-rw-r--r--   3 richard  staff   2561 Jun  9 21:14 README.md
-rw-r--r--   5 richard  staff     31 Jun  9 21:14 edge.d.ts
-rw-r--r--   3 richard  staff     58 Jun  9 21:14 edge.js
drwxr-xr-x   3 richard  staff     96 Jun  9 23:46 generator-build
-rw-r--r--   3 richard  staff     80 Jun  9 21:14 index-browser.js
-rw-r--r--   5 richard  staff     31 Jun  9 21:14 index.d.ts
-rw-r--r--   3 richard  staff     59 Jun  9 21:14 index.js
drwxr-xr-x   3 richard  staff     96 Jun  9 23:46 node_modules
-rw-r--r--   2 richard  staff   6841 Jun  9 21:14 package.json
drwxr-xr-x   7 richard  staff    224 Jun  9 23:46 runtime
drwxr-xr-x  11 richard  staff    352 Jun  9 23:46 scripts
```

If I go to the `src/layers/prisma` directory and run `rm -rf node_modules` and then `npm i` then the packages get installed and prisma client is generated to `./node_modules/@prisma/client` which has:
```
richard:bison prisma > l ./node_modules/@prisma/client
total 88
-rw-r--r--   1 richard  staff  11357 Oct 26  1985 LICENSE
-rw-r--r--   1 richard  staff   2561 Oct 26  1985 README.md
-rw-r--r--   1 richard  staff     31 Oct 26  1985 edge.d.ts
-rw-r--r--   1 richard  staff     58 Oct 26  1985 edge.js
drwxr-xr-x   3 richard  staff     96 Jun 10 00:00 generator-build
-rw-r--r--   1 richard  staff     80 Oct 26  1985 index-browser.js
-rw-r--r--   1 richard  staff     31 Oct 26  1985 index.d.ts
-rw-r--r--   1 richard  staff     59 Oct 26  1985 index.js
-rw-r--r--   1 richard  staff   5279 Jun 10 00:00 package.json
drwxr-xr-x   7 richard  staff    224 Jun 10 00:00 runtime
drwxr-xr-x  11 richard  staff    352 Jun 10 00:00 scripts
```
 and in my IDE the `PrismaClient` is imported correctly.


This is the first time I'm using `pnpm` so I'm not 100% sure I have it setup correctly with the `pnpm-workspace.yaml` file.

**Comment by scottned** [comment]
I was able to resolve this problem by adding this line to my tsconfig.json:
```js
  "include": ["node_modules/.prisma/client/*"],
```

**Comment by richardwardza** [comment]
hhm - I don't get a `node_modules/.prisma` directory generated.

**Comment by richardwardza** [comment]
`prisma generate` creates it in `./../../../node_modules/.pnpm/@prisma+client@3.15.1_prisma@3.15.1/node_modules/@prisma/client`

```
Generated Prisma Client (3.15.1 | library) to ./../../../node_modules/.pnpm/@prisma+client@3.15.1_prisma@3.15.1/node_modules/@prisma/client
```
but nothing in node_modules, only:
```
drwxr-xr-x  22 richard  staff  704 Jun 10 17:33 .
drwxr-xr-x  15 richard  staff  480 Jun 10 17:24 ..
drwxr-xr-x  12 richard  staff  384 Jun 10 17:24 .bin
drwxr-xr-x   3 richard  staff   96 Jun 10 17:24 @prisma
drwxr-xr-x   4 richard  staff  128 Jun 10 17:24 @types
drwxr-xr-x   4 richard  staff  128 Jun 10 17:24 @typescript-eslint
lrwxr-xr-x   1 richard  staff   67 Jun 10 17:24 esbuild -> ../../../../node_modules/.pnpm/esbuild@0.14.43/node_modules/esbuild
lrwxr-xr-x   1 richard  staff   64 Jun 10 17:24 eslint -> ../../../../node_modules/.pnpm/eslint@8.17.0/node_modules/eslint
lrwxr-xr-x   1 richard  staff  129 Jun 10 17:24 eslint-config-airbnb-base -> ../../../../node_modules/.pnpm/eslint-config-airbnb-base@15.0.0_3yxiwxzsqipdmy4jwrlv6vgfmy/node_modules/eslint-config-airbnb-base
lrwxr-xr-x   1 richard  staff  109 Jun 10 17:24 eslint-config-prettier -> ../../../../node_modules/.pnpm/eslint-config-prettier@8.5.0_eslint@8.17.0/node_modules/eslint-config-prettier
lrwxr-xr-x   1 richard  staff  123 Jun 10 17:24 eslint-config-standard -> ../../../../node_modules/.pnpm/eslint-config-standard@17.0.0_zjvxihbqpc3or6hnh25vy3fbci/node_modules/eslint-config-standard
lrwxr-xr-x   1 richard  staff  119 Jun 10 17:24 eslint-plugin-import -> ../../../../node_modules/.pnpm/eslint-plugin-import@2.26.0_pv5w3e62ssxduf5aiwxbc3knra/node_modules/eslint-plugin-import
lrwxr-xr-x   1 richard  staff  115 Jun 10 17:24 eslint-plugin-jest -> ../../../../node_modules/.pnpm/eslint-plugin-jest@26.5.3_zok5b7ewqm4c2qplh4jsnnnkk4/node_modules/eslint-plugin-jest
lrwxr-xr-x   1 richard  staff  102 Jun 10 17:24 eslint-plugin-node -> ../../../../node_modules/.pnpm/eslint-plugin-node@11.1.0_eslint@8.17.0/node_modules/eslint-plugin-node
lrwxr-xr-x   1 richard  staff  122 Jun 10 17:24 eslint-plugin-prettier -> ../../../../node_modules/.pnpm/eslint-plugin-prettier@4.0.0_ucegkljdju7q4zmvwxzqoprf3y/node_modules/eslint-plugin-prettier
lrwxr-xr-x   1 richard  staff  107 Jun 10 17:24 eslint-plugin-promise -> ../../../../node_modules/.pnpm/eslint-plugin-promise@6.0.0_eslint@8.17.0/node_modules/eslint-plugin-promise
lrwxr-xr-x   1 richard  staff   80 Jun 10 17:24 jest -> ../../../../node_modules/.pnpm/jest@28.1.1_@types+node@17.0.41/node_modules/jest
lrwxr-xr-x   1 richard  staff   67 Jun 10 17:24 prettier -> ../../../../node_modules/.pnpm/prettier@2.6.2/node_modules/prettier
lrwxr-xr-x   1 richard  staff   64 Jun 10 17:24 prisma -> ../../../../node_modules/.pnpm/prisma@3.15.1/node_modules/prisma
lrwxr-xr-x   1 richard  staff   93 Jun 10 17:24 ts-jest -> ../../../../node_modules/.pnpm/ts-jest@28.0.4_hhjivpbef6nbltei64cv7t7zz4/node_modules/ts-jest
lrwxr-xr-x   1 richard  staff   86 Jun 10 17:24 ts-loader -> ../../../../node_modules/.pnpm/ts-loader@9.3.0_typescript@4.7.3/node_modules/ts-loader
lrwxr-xr-x   1 richard  staff   71 Jun 10 17:24 typescript -> ../../../../node_modules/.pnpm/typescript@4.7.3/node_modules/typescript
```

**Comment by Jolg42** [comment]
I can reproduce on your project https://github.com/richardwardza/prisma-lambda-layers @richardwardza 

<img width="833" alt="Screen Shot 2022-06-10 at 18 28 22" src="https://user-images.githubusercontent.com/1328733/173110562-259cbdbf-aa71-4273-b281-ad37fe176a66.png">

and if I follow it
<img width="914" alt="Screen Shot 2022-06-10 at 18 28 35" src="https://user-images.githubusercontent.com/1328733/173110650-e5573cb6-de6f-4f46-ad4a-4e32b00ecbc7.png">

**Comment by arthurgailes** [comment]
> 
> Also, in my case, I had to change my import to (note, your relative path may vary):
> 
> ```js
> import { PrismaClient } from '../../node_modules/.prisma/client'
> ```
> 
> Now I can build and run without errors.

This did it for me.

**Comment by richardwardza** [comment]
Thanks @scottned - adding `output = "../node_modules/.prisma/client"` also solves it for me.

**Comment by janpio** [comment]
Note that this optimally should not be neccessary - so we are still treating this as a bug to fix. But good you found a temporary workaround!

**Comment by kasir-barati** [comment]
@Jolg42 Here is what you wanna see:

![image](https://user-images.githubusercontent.com/73785723/173822926-4ea75355-5521-432c-b05b-42bb05bd8b98.png)


It is weird and annoying. What can I do?

**Comment by kasir-barati** [comment]
Really weird. Now it is gone. I guess I am gonna get crazy after a short time if saw such behaviors too much.
I used this:
```ts
import { PrismaClient } from '@prisma/client';
```

I have No idea why it resolved. I did nothing basically. But my linux distro is Manjaro this time. maybe Ubuntu is causing this problem.

**Comment by Joshuajrodrigues** [comment]
> Really weird. Now it is gone. I guess I am gonna get crazy after a short time if see such behaviors too much. I used this:
> 
> ```ts
> import { PrismaClient } from '@prisma/client';
> ```
> 
> I have No idea why it resolved. I did nothing basically. But my linux distro is Manjaro this time. maybe Ubuntu is causing this problem.
> 
> But now `import { Prisma } from '@prisma/client';` does not work
> 
> ```
> Module '"@prisma/client"' has no exported member 'Prisma'.ts(2305)
> ```
> 
> Here is what I did:
> 
> ```batchfile
> [kasir@kasir take-report]$ pnpx prisma generate
> .../share/pnpm/store/v3/tmp/dlx-27253    |   +2 +
> Packages are hard linked from the content-addressable store to the virtual store.
>   Content-addressable store is at: /home/kasir/.local/share/pnpm/store/v3
>   Virtual store is at:             ../.local/share/pnpm/store/v3/tmp/dlx-27253/node_modules/.pnpm
> .../share/pnpm/store/v3/tmp/dlx-27253    | Progress: resolved 2, reused 2, downloaded 0, added 2, done
> Prisma schema loaded from prisma/schema.prisma
> Error: Generator at prisma-dbml-generator could not start:
> 
> /bin/sh: line 1: prisma-dbml-generator: command not found
> 
>  ERROR  Command failed with exit code 1: /home/kasir/.local/share/pnpm/store/v3/tmp/dlx-27253/node_modules/.bin/prisma generate
> ```
> 
> Basically this code works with npm.

thanks

**Comment by lantos-lgtm** [comment]
> It turned out that my previous suggestion to modify tsconfig.json didn't work. However, the suggestion on this [other issue](https://github.com/prisma/prisma/issues/2584#issuecomment-646414549) to add the following to schema.prisma did help:
> 
> ```js
> generator client {
>   provider = "prisma-client-js"
>   output = "../node_modules/.prisma/client"
> }
> ```
> 
> Also, in my case, I had to change my import to (note, your relative path may vary):
> 
> ```js
> import { PrismaClient } from '../../node_modules/.prisma/client'
> ```
> 
> Now I can build and run without errors.


Solved for me, I'm using pnpm.
Could we please put a notification on the docs for pnpm users?
https://www.prisma.io/docs/concepts/components/prisma-client

**Comment by janpio** [comment]
Not really, because this should not even be necessary. Can you please open a new issue with a full description and reproduction what is happening for you? Then we can take a look and fix this properly. (Although of course we are happy you found a working temporary workaround that unblocks you.)

**Comment by arnabtarwani** [comment]
It turns out, using `preserveSymlinks: true` in tsconfig throws this error for me; It tries to import the client from the node_modules directory in the package rather than the root node_modules.

**Comment by qlniu** [comment]
```
generator client {
  provider = "prisma-client-js"
  output = "../node_modules/.prisma/client"
}
```
```
import { PrismaClient } from '../../node_modules/.prisma/client'
```
didn't solve for me, 
env:
```
node      v16.0.0
npm       7.20.5

"@prisma/client": "^4.3.0",
"prisma": "^4.2.1",
```
how to solve?

**Comment by marsidev** [comment]
```prisma
generator client {
  provider = "prisma-client-js"
  output   = "../node_modules/.prisma/client"
}
```

```js
import { PrismaClient } from '.prisma/client'
```

**Comment by mihaic195** [comment]
None of the provided solutions worked for me

**Comment by andrewmclagan** [comment]
Also seeing this on:
```
pnpm@7.11.0
prisma@^4.3.1
```

with a monorepo setup.

**Comment by arnabtarwani** [comment]
With pnpm, the packages are symlinked to avoid multiple dependency issues as you can see here: https://pnpm.io/symlinked-node-modules-structure

When prisma generates a new client, it creates the generated client structure under .pnpm/@prisma/client_(some_id)/node_modules/.prisma/client.

If you want to access this globally you could create .npmrc that will be in the root folder of your monorepo and add the following:

```
strict-peer-dependencies=false
public-hoist-pattern[]=*prisma*
```

This will make the dependency globally available to all the workspaces in the monorepo.

**Comment by nick-fields** [comment]
Encountered this in a yarn (v1.22.18) monorepo where the Prisma schema is in a non-default location.  I worked around it by adding a `prepare` (post-install) script to the root package.json:

```json
"scripts": {
    ...,
    "db:generate-types": "prisma generate --schema=./common/storage/prisma/schema.prisma",
    "prepare": "patch-package && yarn db:generate-types"
},
```

**Comment by irg1008** [comment]
Still no luck with pnpm, I am about to drop pnpm as a package manager, so many flaws

**Comment by muhaimin-anchorzero** [comment]
I'm using pnpm and had luck by specifying a custom location per https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/generating-prisma-client#using-a-custom-output-path.

`packages/prisma/schema.prisma`:
```prisma
generator client {
  provider = "prisma-client-js"
  output   = "prisma-generated/client"
}
```

`packages/prisma/src/index.ts`:
```ts
export * from '../prisma-generated/client';
```

`packages/prisma/package.json`:
```
...
    "name": "@internal/prisma",
...
```

`packages/app/src/index.ts`:
```ts
import { PrismaClient } from '@internal/prisma';
```

**Comment by AlaaZorkane** [comment]
Worked for me by setting `"preserveSymlinks": false` in `tsconfig`

**Comment by ralcorta** [comment]
> With pnpm, the packages are symlinked to avoid multiple dependency issues as you can see here: https://pnpm.io/symlinked-node-modules-structure
> 
> When prisma generates a new client, it creates the generated client structure under root/node_modules/.pnpm/@prisma/client_(some_id)/node_modules/.prisma/client.
> 
> If you want to access this globally you could create .npmrc that will be in the root folder of your monorepo and add the following:
> 
> ```
> strict-peer-dependencies=false
> public-hoist-pattern[]=*prisma*
> ```
> 
> This will make the dependency globally available to all the workspaces in the monorepo.

This solved my errors when importing generated models from Primsa 🚀

**Comment by Atomic71** [comment]
I've been at this for two days now, keep thinking this is somehow solve-able.

I have a turborepo, deployed on vercel, and am trying to use Prisma as a standalone package so I can import it into my two native apps which rely on the said monorepo. 

The crazy thing is, in local all is well and good, so there's not a good way to debug this.

**Comment by ctdh** [comment]
apparently restarting TypeScript resolves this...
https://stackoverflow.com/questions/69274503/property-does-not-exist-when-i-want-to-use-model-added-in-prisma-schema

**Comment by pnxdxt** [comment]
I am running into this issue during our CI w/ GitHub actions

**Comment by timmeade** [comment]
```
generator client {
    provider = "prisma-client-js"
    output = "../node_modules/.prisma/client"
}
```

Fixed for me.

**Comment by phifa** [comment]
huge issues with turborepo, pnpm and prisma. tried it all, nothing works reliably.

**Comment by AdditionAddict** [comment]
Faced the same issue in an Nx monorepo using pnpm v8.1.0, node v16.17.1, windows.

Solved by having a `npx prisma generate` target called `"prisma-generate"` and for each app and library where `@prisma/client` is used, putting `"dependsOn": [ "prisma-generate" ]` within the `"build"` target. The hint here was it was only failing for my GitHub CI where the library was built in isolation. 

Tagging @pnxdxt in case above is useful

**Comment by craigcosmo** [comment]
install node 16 will sovle, prisma 4 doesn't support certain version of node.

**Comment by imCorfitz** [comment]
For me, it was due to VSCode apparently "caching" the directories. I had to `Reload Window` and then everything was imported correctly without errors.

**Comment by millsp** [comment]
@scottned @lantos1618 @phifa @ralcorta @andrewmclagan @timmeade @irg1008 

1. Are you by any chance using `preserveSymlinks`?
I confirmed that this is causing issues with Prisma Client  and `pnpm`, setting it to `false` prevents type issues.

2. Have you tried to restart VSCode or the TS Server?
Some users have reported that after being generate, VSCode did not pick up the new generated types.

3. Have you tried manually running `prisma generate`, did that change anything?
Some users have reported that Prisma Client auto-generate does not work well on `pnpm`.

There are many issues interleaved, and so far we know my three mentions just above are things we need to fix.

Did any of these help? If not, could you please provide us with a small reproduction? We are not able to reproduce the issues mentioned in the comments, so we need your help :pray:  I'd appreciate if you could submit us some code.

**Comment by lantos1618** [comment]
It is probably the first point. PreserveSymLinks sounds like the one to do the job but what I am doing now is just using this 

// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["jsonProtocol"]
  output          = "../src/generated/prisma/client"
}

and  adding src/generated to .gitignore 

This way if someone clones the repo it will work for them without them needing to change settings if they use pnpm

**Comment by ershisan99** [comment]
For me the problem happened after running `pnpx prisma generate`, what helped was to update prisma to the latest version in project deps, run `pnpm prisma generate` (not pnpx) and restart ts service. Hope this helps someone.

**Comment by nixjs** [comment]
I deployed on vercel and have the same issue.  How to fix it please?
<img width="1261" alt="image" src="https://github.com/prisma/prisma/assets/23132483/2fabc4e4-fa92-43e2-b9c8-6e735c4489ea">

**Comment by millsp** [comment]
@nixjs Your error is slightly different, it says that `SocialProviderLogin` cannot be imported. Does this also happen with `@prisma/client`?

**Comment by nixjs** [comment]
> I deployed on vercel and have the same issue. How to fix it please? <img alt="image" width="1261" src="https://user-images.githubusercontent.com/23132483/254428132-2fabc4e4-fa92-43e2-b9c8-6e735c4489ea.png">

I resolved

**Comment by eakl** [comment]
I have strictly followed [this tutorial](https://turbo.build/repo/docs/handbook/tools/prisma) and this is the error I get:

```bash
Unable to interop `export * from "@prisma/client"` in /Users/me/project-root/packages/database/index.ts, this may lose module exports. Please export "@prisma/client" as ESM or use named exports instead, e.g. `export { A, B } from "@prisma/client"`
```

I have also tried the fixes proposed above without success.

The issue is one year old but still getting impacted people. Monorepo is not a edge.

**Comment by ctjlewis** [comment]
This package is not specified correctly. It is an error.

**Comment by ctjlewis** [comment]
### Solution

Run **`pnpm prisma generate`** to create the `.prisma/client` folder that `@prisma/client` attempts to import from.

**Comment by karis** [comment]
it works for me,

i also using:
pnpm v8
nuxt 3.6.5
prisma 5. and read the update. i already upgrade to 5.1.1
`pnpm upgrade prisma@5 @prisma/client@5`
IDE VSCode

[references](https://www.prisma.io/docs/concepts/components/prisma-client)
`
const { PrismaClient } = require('@prisma/client')
const prisma = new PrismaClient()
// use `prisma` in your application to read and write data in your DB
`

**Comment by sxy15** [comment]
npx prisma generate

**Comment by kyle-belle** [comment]
> I have strictly followed [this tutorial](https://turbo.build/repo/docs/handbook/tools/prisma) and this is the error I get:
> 
> ```shell
> Unable to interop `export * from "@prisma/client"` in /Users/me/project-root/packages/database/index.ts, this may lose module exports. Please export "@prisma/client" as ESM or use named exports instead, e.g. `export { A, B } from "@prisma/client"`
> ```
> 
> I have also tried the fixes proposed above without success.
> 
> The issue is one year old but still getting impacted people. Monorepo is not a edge.
> 
> **EDIT**
> 
> VSCode tries to import from `project-root/node_modules/.pnpm/@prisma+client@5.0.0_prisma@5.0.0/node_modules/@prisma/client` while the generated path is `project-root/node_modules/.pnpm/@prisma+client@5.0.0_prisma@5.0.0/@prisma/client` (without `node_modules`)

Did you ever manage to solve this. I too am trying to get a mono-repo up and run where i am able to import prisma types and enums on the front end

**Comment by kyle-belle** [comment]
> > I have strictly followed [this tutorial](https://turbo.build/repo/docs/handbook/tools/prisma) and this is the error I get:
> > ```
> > Unable to interop `export * from "@prisma/client"` in /Users/me/project-root/packages/database/index.ts, this may lose module exports. Please export "@prisma/client" as ESM or use named exports instead, e.g. `export { A, B } from "@prisma/client"`
> > ```
> > 
> > 
> > I have also tried the fixes proposed above without success.
> > The issue is one year old but still getting impacted people. Monorepo is not a edge.
> > **EDIT**
> > VSCode tries to import from `project-root/node_modules/.pnpm/@prisma+client@5.0.0_prisma@5.0.0/node_modules/@prisma/client` while the generated path is `project-root/node_modules/.pnpm/@prisma+client@5.0.0_prisma@5.0.0/@prisma/client` (without `node_modules`)
> 
> Did you ever manage to solve this. I too am trying to get a mono-repo up and run where i am able to import prisma types and enums on the front end


For some extra details:

The frontend uses VITE

This is the console error: `Uncaught SyntaxError: ambiguous indirect export: UserType`

This is the line when i view the transpiled source the browser is trying to run `import { UserType } from "/@fs/home/[user]/Documents/[project]/packages/db/index.ts";`

**Comment by kyle-belle** [comment]
> I have strictly followed [this tutorial](https://turbo.build/repo/docs/handbook/tools/prisma) and this is the error I get:
> 
> ```shell
> Unable to interop `export * from "@prisma/client"` in /Users/me/project-root/packages/database/index.ts, this may lose module exports. Please export "@prisma/client" as ESM or use named exports instead, e.g. `export { A, B } from "@prisma/client"`
> ```
> 
> I have also tried the fixes proposed above without success.
> 
> The issue is one year old but still getting impacted people. Monorepo is not a edge.
> 
> **EDIT**
> 
> VSCode tries to import from `project-root/node_modules/.pnpm/@prisma+client@5.0.0_prisma@5.0.0/node_modules/@prisma/client` while the generated path is `project-root/node_modules/.pnpm/@prisma+client@5.0.0_prisma@5.0.0/@prisma/client` (without `node_modules`)

@eakl 

If all you want to do is use the enums on the frontend you can create a new in the database package like `/Users/me/project-root/packages/database/prisma/client/index.ts`

with contents:
```typescript
export {$Enums} from '@prisma/client';
```

and be sure to change all front-end imports to use this new file

for the model types i suggest using something like [zod-prisma-types](https://www.npmjs.com/package/zod-prisma-types)

and have it generate types in a folder like `/Users/me/project-root/packages/database/prisma/generated/` (make sure its git ignored btw)

and on the front-end import all model types from there. This will require installing [zod](https://zod.dev/) if you don't already use it but i highly recommend it

**Comment by SilasRodrigues19** [comment]
> ### Solution
> Run **`pnpm prisma generate`** to create the `.prisma/client` folder that `@prisma/client` attempts to import from.

Thanks @ctjlewis 

It worked

**Comment by leafleaf90** [comment]
> For me the problem happened after running `pnpx prisma generate`, what helped was to update prisma to the latest version in project deps, run `pnpm prisma generate` (not pnpx) and restart ts service. Hope this helps someone.

`pnpm prisma generate` solved it for me too.

**Comment by Frostbourn** [comment]
In my scenario within a monorepo setup, the solution was simple. I just needed to include `"postinstall": "prisma generate"` in the `package.json` file. Additionally, importing the Prisma client was straightforward: `import { PrismaClient } from "@prisma/client/edge";`

**Comment by MrEmanuel** [comment]
I also had this problem after upgrading some packages, both prisma, pnpm and vercel. 
I solved it by reloading vscode window `Reload Window`

Annoying.

**Comment by millsp** [comment]
@MrEmanuel which OS are you running on? (This info might be relevant because of our VSCode extension).

**Comment by MrEmanuel** [comment]
@millsp Mac OS. Sonoma 14.2.1

**Comment by Material-X** [comment]
I'm getting an error similar to that, but the error suggest that @client/prisma has no exported memeber "OrderItemCreateInput". I was expecting to have this import after running the npx prisma generate command. 

![image](https://github.com/prisma/prisma/assets/58167461/9a2b2b3b-eba6-4d1f-94ef-3df9e7ccd7f2)

**Comment by Sibindra** [comment]
I found a solution that worked for me. i removed the current installation of Prisma Client and generated it on init as per the documentation:

```sh

npx prisma migrate dev --name init
```

[link to documentation](https://www.prisma.io/docs/getting-started/setup-prisma/start-from-scratch/relational-databases/using-prisma-migrate-typescript-postgresql)

this resolved the issue for me. I encourage anyone facing similar problems to try this approach

feel free to dive deeper into the root cause of the issue and provide further insights

**Comment by burakcan** [comment]
Today I suddenly started to have this issue by just triggering a deploy on Vercel. Nothing changed in the codebase. Now it fails both on my local and on Vercel - including the old builds that succeeded at the time.

**Comment by dominics** [comment]
When this happened to me (pnpm monorepo with a `public-hoist-pattern`):
 - The `prisma generate` call in the `postinstall` from `./some/package` was detecting pnpm and reporting:
   > `✔ Generated Prisma Client (v5.11.0) to ./../../node_modules/.pnpm/@prisma+client@5.11.0_prisma@5.11.0/node_modules/@prisma/client in 89ms`
   - This is the correct relative path to PNPM's source directory, equivalent to `node_modules/.pnpm/@prisma+client@5.11.0_prisma@5.11.0/node_modules/@prisma/client` from the repository root
 - However, it actually generated the `.prisma/` directory at `node_modules/.pnpm/@prisma+client@5.11.0_prisma@5.11.0/node_modules/.prisma`
    - For the `require('.prisma/client/default')` call in `node_modules/.pnpm/@prisma+client@5.11.0_prisma@5.11.0/node_modules/@prisma/client/default.js` to find this file it would need to check in it's _parent directory's node_modules dir_ (i.e. following the standard [resolution algo](https://nodejs.org/api/modules.html#loading-from-node_modules-folders))
    - However, pnpm _doesn't_ seem to do this (and it may be unfair to ask it to, within the `.pnpm/` file hierarchy, which as I understand it is private implementation details of pnpm - it's not like these are the final node_modules that get patched beside a package.json file, after all)
 - If the `.prisma/` directory is instead generated at `node_modules/.pnpm/@prisma+client@5.11.0_prisma@5.11.0/node_modules/@prisma/client/node_modules` (i.e. _inside_ the `node_moules/` of `@prisma/client` not just adjacent to it), then it is found by `default.js` and everything works from there

**Comment by kilimanjjjaro** [comment]
> I found a solution that worked for me. i removed the current installation of Prisma Client and generated it on init as per the documentation:
> 
> ```shell
> npx prisma migrate dev --name init
> ```
> 
> [link to documentation](https://www.prisma.io/docs/getting-started/setup-prisma/start-from-scratch/relational-databases/using-prisma-migrate-typescript-postgresql)
> 
> this resolved the issue for me. I encourage anyone facing similar problems to try this approach
> 
> feel free to dive deeper into the root cause of the issue and provide further insights

This worked for me, thanks!

**Comment by ortonomy** [comment]
> I'm getting an error similar to that, but the error suggest that @client/prisma has no exported memeber "OrderItemCreateInput". I was expecting to have this import after running the npx prisma generate command.
> 
> ![image](https://private-user-images.githubusercontent.com/58167461/313720361-9a2b2b3b-eba6-4d1f-94ef-3df9e7ccd7f2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU0NDIwMzcsIm5iZiI6MTcxNTQ0MTczNywicGF0aCI6Ii81ODE2NzQ2MS8zMTM3MjAzNjEtOWEyYjJiM2ItZWJhNi00ZDFmLTk0ZWYtM2RmOWU3Y2NkN2YyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTExVDE1MzUzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMzYzA3MWQwZWZlMTJkMDkzZWRmZjNlMzNhNjM5YjI3ZjUzMWZhZjE3N2E4MGRhZDlhNTFiOGM3YTBlMTExMzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.aNBIFyqy-DsZVp5AMoewfKgNpFvvTaFEid4krSwV6oU)

Also seeing this. I can't resolve it

**Comment by abitocodes** [comment]
I typically use pnpm for my development work, so encountering this issue has been quite frustrating.

In my experience, the error was resolved after running `pnpm install` followed by `pnpm dlx prisma generate`. I'm not very familiar with how Prisma functions, so it's possible that this solution might just be a coincidence.

**Comment by Dagwal** [comment]
For me it works installing only the @nestjs/prisma but not both @nestjs/prisma and prisma
![image](https://github.com/prisma/prisma/assets/108694236/8e1ee57c-1ffd-4cca-aa2d-b8cbe7b0ea18)

**Comment by hafaiedhmehdy** [comment]
It's a **_bit_** drastic, so first, make a copy of your repository to ensure you don't ruin your project. However, running the following command usually fixes this type of issue:

```bash
rm -rf node_modules pnpm-lock.yaml && pnpm install
```

This just happened to me today, and it worked great. 😎

**Comment by eyoba-bisru** [comment]
<code>npx prisma migrate dev</code> OR <code>npx prisma generate</code> solved for me

**Comment by mengnans** [comment]
I am uxing NX monorepo with pnpm & prisma

This helped my case, instead of generating the .prisma/client under the node_modules, I changed the path to node_modules/@prisma/client/node_modules/.prisma/client

```
generator client {
  provider        = "prisma-client-js"
  output          = "../../../node_modules/@prisma/client/node_modules/.prisma/client"
  binaryTargets   = ["native", "debian-openssl-3.0.x"]
  previewFeatures = ["views"]
}
```

**Comment by Chatelo** [comment]
I added this in my package.json and it fixed
`{
  ...
  "scripts" {
    "postinstall": "prisma generate"
  }
  ...
}`

and in schema is add outpu path as follow:
`generator client {
  provider = "prisma-client-js"
  output = "../../node_modules/.prisma/client"
}`
and lastly in my lib/prisma.ts
`import PrismaClient from "@prisma/client";

export const prisma = new PrismaClient.PrismaClient();
export default prisma;`

and **BOOM!** the problem is fixed

**Comment by MuntasirSZN** [comment]
> I added this in my package.json and it fixed
> 
> ```ts
> {
>   ...
>   "scripts" {
>     "postinstall": "prisma generate"
>   }
>   ...
> }
> ```
> 
> and in schema add output path as follow:
> 
> ```ts
>  generator client {
>   provider = "prisma-client-js"
>   output = "../../node_modules/.prisma/client"
> }
> ```
> 
> and lastly in my lib/prisma.ts
> 
> ```ts
>  import PrismaClient from "@prisma/client";
> 
> export const prisma = new PrismaClient.PrismaClient();
> export default prisma;
> ```
> 
> Finally run `npx prisma generate ` You might also want to reload your VSCODE
> 
> and **BOOM!** the problem is fixed

Thanks a lot.

**Comment by ee99ee** [comment]
Contrary to what almost everyone said about this issue, I _prepended_ `prisma generate` to my build step before running in my case `next build`, and that worked for me:

```
{
  ...
  "scripts" {
    "build": "prisma generate && next build"
  }
  ...
}
```

**Comment by mz-dfhp** [comment]
The problem is still unresolved

**Comment by DobromirKirovLime** [comment]
Is this ever going to be resolved :(

**Comment by JohnatanPalacios** [comment]
I encountered an issue where @prisma/client was not found. In my prisma/schema.prisma, I had a User model with an id using cuid(), along with createdAt and updatedAt timestamps. The solution was to install the necessary dependencies and regenerate the Prisma client. I ran the following commands:

```
pnpm install @prisma/client  
pnpm install prisma --save-dev  
npx prisma generate  
```

After running these, the issue was resolved.

**Comment by Joshuajrodrigues** [comment]
I was using prisma for a hobby project when I found this issue.
I've been subscribed to it since then. I work a full time job now.  I just follow this thread as curiosity.

**Comment by romeonombrejr** [comment]
> generate

This worked like magic! Thanks!

**Comment by rubkn** [comment]
I also ran into this issue while using `pnpm`. Switching to `yarn` instead of `pnpm` fixed it for me. Not sure if it’s related to how `pnpm` handles dependencies, but just wanted to share in case it helps others.

**Comment by romeonombrejr** [comment]
> I added this in my package.json and it fixed
> 
> {
>   ...
>   "scripts" {
>     "postinstall": "prisma generate"
>   }
>   ...
> }
> and in schema add output path as follow:
> 
>  generator client {
>   provider = "prisma-client-js"
>   output = "../../node_modules/.prisma/client"
> }
> and lastly in my lib/prisma.ts
> 
>  import PrismaClient from "@prisma/client";
> 
> export const prisma = new PrismaClient.PrismaClient();
> export default prisma;
> Finally run `npx prisma generate ` You might also want to reload your VSCODE
> 
> and **BOOM!** the problem is fixed

This worked like magic! Thanks!

**Comment by Adam21e** [comment]
For anyone hitting this issue that is using pnpm v10, I had to run `pnpm approve-builds` and approving the relevant builds as I missed the warning from pnpm about ignored build scripts on initial install.

For future reference you can approve build scripts when you install with `pnpm add <package-name> --allow-build=<package-name>`  https://pnpm.io/cli/add#--allow-build

**Comment by romeonombrejr** [comment]
> > I added this in my package.json and it fixed
> > {
> > ...
> > "scripts" {
> > "postinstall": "prisma generate"
> > }
> > ...
> > }
> > and in schema add output path as follow:
> > generator client {
> > provider = "prisma-client-js"
> > output = "../../node_modules/.prisma/client"
> > }
> > and lastly in my lib/prisma.ts
> > import PrismaClient from "@prisma/client";
> > export const prisma = new PrismaClient.PrismaClient();
> > export default prisma;
> > Finally run `npx prisma generate ` You might also want to reload your VSCODE
> > and **BOOM!** the problem is fixed
> 
> This worked like magic! Thanks!

Hi! This has been a recurring issue for me and somehow this fix above is not a consistent solution, especially if you're working with a project from multiple machines. But hey! I think I found the best solution for this one.


1. It's starts with how you generate your schema.prisma with the command "npx prisma init --db --output ../generated/prisma". Similar to how it should be in the documentation:

https://www.prisma.io/docs/getting-started/setup-prisma/start-from-scratch/relational-databases-typescript-prismaPostgres


2. Notice that the output location "/generated/prisma", this should be similar to the output directory in your schema.prisma file:

generator client {
  provider = "prisma-client-js"
  output = "../generated/prisma"
}


3. Now proceed with migrating your schema,prisma file with the command "npx prisma migrate dev --name init".


4. Finally when using PrismaClient, the 'from' should be the location of the output files, which is 'generated/prisma'. Like this:
import { PrismaClient } from 'generated/prisma'; 


And..... That's it! Hopefully this works for you as well!

**Comment by Nucletic** [comment]
thanks brother just changing the output and running `npx prisma generate` solved my problem: 

prisma.ts

`generator client {
  provider = "prisma-client-js"
  output   = "../node_modules/.prisma/client" // after
  // output   = "../src/generated/prisma" before
}`

**Comment by iamgutz** [comment]
The following worked for me:

```js
// package.json
"scripts": {
//...
"postinstall": "npx prisma generate"
}
```

```js
// schema.prisma
generator client {
   provider = "prisma-client-js"
   output = "../../node_modules/.prisma/client"
}

datasource db {
    provider = "postgresql"
    url = env("POSTGRES_URL")
}
//...
```

**Comment by ussfranck** [comment]
Since prisma version ^6.6.0, the way in which model client types are generated has changed. We don't know why my prisma declares the output of TS types to be **"output = "../lib/generated/prisma""** but it's **obviously wrong**, it's rather in **"node_modules/.prisma/client"**, so in the **"schema.prisma"** file the client should be:

```prisma
generator client {
    provider = "prisma-client-js"
    + output = "../../node_modules/.prisma/client"  // Set correct path
    - output = ". ./lib/generated/prisma"  // Remove this it
}
```

You may need to refine the entire dependency installation, especially if you are running pnpm, remove the **"pnpm-lock.yaml"** as **PNPM symbolic links** can maintain a strange cache which can maintain the problem, also remove the **"node_modules"**, and finally restart your IDE.

**Comment by feliipesouz** [comment]
O erro acontece quando o Prisma Client:

Não foi gerado corretamente ou foi configurado com output customizado e o import não foi ajustado.

- Remova o output customizado no schema.prisma:
generator client {
  provider = "prisma-client-js"
  // output removido!
}

npx prisma generate

import { PrismaClient } from '@prisma/client'

- se o erro persistir, é bom limpar e reinstalar os módulos:

rm -rf node_modules
rm -rf pnpm-lock.yaml
pnpm install
npx prisma generate

**Comment by ghbishal** [comment]
since prisma version ^6.6.0, there is a warning that we need to specify the output for generated,

since i have my generated output in this way

``` prisma
generator client {
  provider = "prisma-client-js"
  output   = "../src/generated/prisma"
}
```

so my prisma.ts should also be like this.

```ts
import { PrismaClient } from '@/generated/prisma';

const prisma = new PrismaClient();

const globalForPrisma = global as unknown as { prisma: typeof prisma };

if (process.env.NODE_ENV !== 'production') {
  globalForPrisma.prisma = prisma;
}

export default prisma;

```

then in package.json you have to add this line

```json
    "build": "prisma generate && next build",
```

This should pass the build.
**`PS DO not forget to add the generate folder in .gitignore`**
in my case (similarly do same in your case.)
```bash
/src/generated/prisma
```

**Comment by ivanmayta** [comment]
If no output path is specified it works for "@prisma/client" after running "prima generate", but shows the following warning, 

![Image](https://github.com/user-attachments/assets/68e3312e-0d48-4c02-b007-2c318ea6a694)

More details that resolves the issue: https://pris.ly/cli/output-path

**Comment by maximtrp** [comment]
Seems like you can also try the new "prisma-client" generator. I have switched to it when on v6.6.0 and got no errors while building my project with custom client path. Still not sure about its stability.

```
generator client {
  provider = "prisma-client"
  output   = "./client"
}
```

The client is generated to `./prisma/client` directory of the project.

**Comment by Arav-Menon** [comment]
`generator client {
  provider = "prisma-client-js"
  // output   = "../generated/prisma"
}`
**Remove this output from schema.prisam and generate the prisma client again and you are good to go**

**Comment by XiaoPengYouzzZ** [comment]
I changed [import { PrismaClient } from "@prisma/client"] to [import { PrismaClient } from "~/generated/prisma"] and the problem was solved, but I don't know if it's a coincidence

**Comment by aimerthyr** [comment]
if use pnpm , you must delete output, let pnmm to find byself. it works for me 

<img width="654" height="386" alt="Image" src="https://github.com/user-attachments/assets/34a482c2-37f7-48ea-86b5-fdf4730e3dd2" />

**Comment by jy-joung-bitor** [comment]
I used `@prisma/client` in `prisma-client` but the error happened. Is `@prisma/client` not recommended in `prisma-client`?

**Comment by uttampun44** [comment]
you have to import from generated folder 

import { PrismaClient } from '../generated/prisma/client';  

Like this, it will work.
